### PR TITLE
feat: [IDP-1164]: Plugin config reader

### PR DESCRIPTION
**Pull request by vikyathharekal**
## Describe your changes
This is the first cut of the config reader. (to unblock the Pagerduty data source work)
It fetches the app config from DB and evaluates the given expressions to get the value for them.
If the value contains an environment variable, it resolves and returns the decrypted value for it.

### Pagerduty Config
```
proxy:
  /pagerduty:
    target: https://api.pagerduty.com
    pathRewrite:
      api/proxy/pagerduty/?: /
    headers:
      Authorization: "Token token=${MY_PAGERDUTY_TOKEN}"
```

### How to call?
```
configReader.getConfigValues(accountIdentifier, "appConfig.proxy.\"/pagerduty\".target");
```

Note: The support for configs having an array (that is needed for the k8s plugin) will be added in subsequent PR.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52819)

### CODE REVIEW

File: idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    IdpExpressionEvaluator evaluator = new IdpExpressionEvaluator(Map.of(APP_CONFIG_CONTEXT, yamlData));
+    Map<String, Object> keyValuesMap = new HashMap<>();
+    for (String keyExpression : keyExpressions) {
+      Object value = null;
+      try {
+        value = evaluator.evaluateExpression(keyExpression, ExpressionMode.RETURN_NULL_IF_UNRESOLVED);
+      } catch (Exception e) {
+        log.warn("Could not evaluate expression {} while reading config", keyExpression);
+      }
+      if (value == null) {
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java:53]: Can we remove this null check and put continue statement on the catch block itself ?

File: idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        log.warn("Could not find the required data by evaluating expression for {}", keyExpression);
+        continue;
+      }
+      keyValuesMap.put(keyExpression, getDecryptedValueIfNeeded(accountIdentifier, value));
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java:57]: Can be put inside the try block itself

File: idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    for (String keyExpression : keyExpressions) {
+      Object value = null;
+      try {
+        value = evaluator.evaluateExpression(keyExpression, ExpressionMode.RETURN_NULL_IF_UNRESOLVED);
+      } catch (Exception e) {
+        log.warn("Could not evaluate expression {} while reading config", keyExpression);
+      }
+      if (value == null) {
+        log.warn("Could not find the required data by evaluating expression for {}", keyExpression);
+        continue;
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java:55]: So, if are not able to evaluate it, we just continue. Will we not hit problems afterwards ?
Like something is needed on the datasource layer, but we have not got it.

File: idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        throw new UnexpectedException(String.format("Could not get the decrypted value for secret: %s, used "
+                + "by env: %s, in account: %s",
+            secret.getHarnessSecretIdentifier(), envName, accountIdentifier));
+      }
+      value = ((String) value).replace(env, decryptedValue);
+    }
+    return value;
+  }
+
+  public static Map<String, Object> parseYamlString(String yamlString) {
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java:87]: nit: can be moved to YamlUtils in common package

File: idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java (hunk 1 above)
Comment (vikyathharekal) [reply idp-service/src/main/java/io/harness/idp/scorecard/datasources/utils/ConfigReader.java:50]: fixed as discussed offline

**Review by sathish-soundarapandian — APPROVED** [review]

**Review by devesh-harness — DISMISSED** [review]

**Comment by vikyathharekal** [comment]
trigger review

**Comment by vikyathharekal** [comment]
trigger uts

**Review by devesh-harness — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger uts

**Review by sandeepa-kv-jois — APPROVED** [review]

**Comment by vikyathharekal** [comment]
trigger ut0

**Comment by vikyathharekal** [comment]
trigger ut1, ut4

**Comment by vikyathharekal** [comment]
trigger ut0, review

**Comment by vikyathharekal** [comment]
trigger ut1, ut4

**Pull request merged**